### ISSUE TITLE
IAVL panic with missing value

### ISSUE THREAD

**Issue by erikgrinaker** [cosmos/iavl#261]
A user is reporting the following panics with the SDK (0.38.4) and options `pruning=everything`:

```
I[2020-06-08|14:01:38.646] starting ABCI with Tendermint                module=main 
panic: Value missing for hash 190e17e5722f481b33536af9525e21abd9e9a02e8e4fed46e3921010899ab65c corresponding to nodeKey n�r/HSj�R^!���.�O�F����\
goroutine 1 [running]:
github.com/tendermint/iavl.(*nodeDB).GetNode(0xc000df6580, 0xc002a77f60, 0x20, 0x20, 0x0)
        /home/ubuntu/go/pkg/mod/github.com/tendermint/iavl@v0.13.2/nodedb.go:109 +0x7bf
```

This does not happen when pruning is disabled.

We've seen similar panics related to concurrency, see e.g. #240.

**Comment by alexanderbez** [comment]
Turns out the user changed pruning strategies inflight (via restart). While this behavior _should_ be supported, I doubt it's functional and most likely due to the fact that we attempted to prune a version that did not exist (i.e. `prune=nothing` => `prune=everything`).

**Comment by zmanian** [comment]
Based on the latest comment here: https://github.com/cosmos/cosmos-sdk/issues/6370 the problem can be reproduced when `pruning='everything"` is set from genesis.

The expected behavior here is that if pruning is `everything", the ABCI application would replay ever block from genesis against an empty IAVL tree to restore state to the current block height and it would never look for a snapshot.

**Comment by zmanian** [comment]
I have some long running prune=everything nodes from goz phase 3 that I haven't tried to restart yet to see if I can reproduce.

**Comment by erikgrinaker** [comment]
> The expected behavior here is that if pruning is `everything", the ABCI application would replay ever block from genesis against an empty IAVL tree to restore state to the current block height and it would never look for a snapshot.

You mean this is the expected buggy behavior, not the desired behavior, right? With `pruning=everything`, I would expect the application to retain the latest version on disk, and start from that.

**Comment by erikgrinaker** [comment]
I suspect this may be as simple as `DeleteVersion()` being unsafe. Running with `pruning=everything` means that we are deleting a version for every single height, and from what I can tell `DeleteVersion()` can leave incomplete data on unclean shutdown, for two reasons:

* IAVL first deletes nodes, then the version root - this can lead to an incomplete version still being visible.

* GoLevelDB batches are [not written atomically](https://github.com/syndtr/goleveldb/blob/758128399b1df3a87e92df6c26c1d2063da8fabe/leveldb/db_write.go#L271), for small batches (<4 MB by default).

I'll try to verify this - it's a bit unclear exactly how this might lead to the observed crashes. The fixes should be fairly straightforward, but may have performance implications.

**Comment by erikgrinaker** [comment]
Confirmed, I'm seeing this with a default `iavl.NewMutableTree` with GoLevelDB an no IAVL pruning. Deleting the previous version after saving the current (which is the same thing the SDK does for its own IAVL pruning) gives this:

```
Saved version 66
Deleted version 65
^Csignal: interrupt
erik@gMac torture $ go run ./main.go 
Loaded version 66
panic: Value missing for hash 0654c14468757f8c2afec9ab3cb2b38c3de8c612bb865439f0608a9b0dc65eea corresponding to nodeKey 6e0654c14468757f8c2afec9ab3cb2b38c3de8c612bb865439f0608a9b0dc65eea

goroutine 1 [running]:
github.com/tendermint/iavl.(*nodeDB).GetNode(0xc00018e300, 0xc000180f20, 0x20, 0x20, 0x0)
	/Users/erik/Projects/tendermint/iavl/nodedb.go:189 +0x7bf
```

**Comment by erikgrinaker** [comment]
The restarts were a red herring, I've been able to reproduce this by creating and deleting a sequence of versions with specific key/value pairs - added a [test-case](https://github.com/tendermint/iavl/blob/4882c9bbebeac91aab80dac923ce1687cdac2ce1/mutable_tree_test.go#L223) for it. I'm going through the orphan handling to identify the problem, but it's unfamiliar and non-trivial code so it may take a little while.

**Comment by erikgrinaker** [comment]
This turned out to be caused by a subtle flaw in the test case - consider this snippet:

```go
key := make([]byte, 16)
value := make([]byte, 16)
random.Read(key)
random.Read(value)
// if we hit an existing key, regenerate and set a new key
for tree.Set(key, value) {
    random.Read(key)
}
```

Notice that the inner `random.Read(key)` call reads into _the same byte-slice_ as originally used. The `Set()` call will have placed this byte-slice into the IAVL node, and I think it's somehow made its way into the orphans list. So when we change the key, this changes the nodes in the tree (giving them the wrong node position for their key), and I think somehow also changes which orphan nodes are scheduled for removal, causing the node to be missing in the next version.

Changing the loop as follows makes the panic go away:

```go
for tree.Set(key, value) {
    key := make([]byte, 16)
    random.Read(key)
}
```

On one hand I'm relieved that this was a problem with the test rather than IAVL, but on the other hand I'm back to square one for debugging this crash. I would not be surprised if the user crash is caused by a similar problem in the SDK, since this is a very easy mistake to make, and I don't believe this is even documented for IAVL (although it is for tm-db).

As a side-note, `MemDB` in tm-db 0.5.0 also has this behavior (given and returned keys and values are pointers to the actual database items, for performance). In MemDB 0.4.0, as an implementation detail, the key was copied due to a string-cast for use as a map key, while the value was stored as a pointer. It is quite possible that this change has triggered similar bugs, although MemDB should not be involved at all with `pruning=everything` for GoLevelDB-based IAVL trees.

I'm considering further approaches to debugging this.

**Comment by erikgrinaker** [comment]
I've been able to reproduce this again, have written a test-case here:

https://github.com/cosmos/iavl/blob/d554b885042a47ea93d6b998f8b869bab37dfdff/mutable_tree_test.go#L270

The behavior is basically this:

Use `KeepEvery: 5` and `KeepRecent: 1`. When we save a version to disk (every 5), delete the previous saved version (so after 10 delete 5). This follows SDK pruning behavior. Stop producing versions after version 12.

Open a new tree, loading version 10, and write 12 new versions with the same key/value pairs as from the previous run. After version 15 is saved and version 10 is deleted, IAVL panics:

```
panic: Value missing for hash e61c06828e90b91404a9442f74d33f3b89f9d9fa93960df459a7ca313a99338b corresponding to nodeKey 6ee61c06828e90b91404a9442f74d33f3b89f9d9fa93960df459a7ca313a99338b [recovered]
	panic: Value missing for hash e61c06828e90b91404a9442f74d33f3b89f9d9fa93960df459a7ca313a99338b corresponding to nodeKey 6ee61c06828e90b91404a9442f74d33f3b89f9d9fa93960df459a7ca313a99338b
```

I don't think there are any flaws in the test case this time, but would appreciate a second pair of eyes @alexanderbez.

**Comment by erikgrinaker** [comment]
I've found the problem, and written a minimal test case: https://github.com/cosmos/iavl/blob/60acfb1ff36c057882d418b0bbe04aa8d759edfc/mutable_tree_test.go#L336

It boils down to orphan entries being incorrectly persisted to disk for changes that are only done in memory, and then reloading the tree and making _different_ changes causes these orphan entries to remove nodes that are now in use when the versions the were included in are deleted. Consider this history, with keepEvery=2 keepRecent=1:

* Save version 1 (memory): set a=1 b=2
* Save version 2 (disk): set c=3 d=4
* Save version 3 (memory): set d=0 → orphan d=4 for version > 2 (persisted to disk)
* Restart, load version 2 from disk
* Save version 3 (memory): set a=0
* Save version 4(disk): set b=0
* Delete version 2 (disk) → orphan for d=4 is no longer needed, so node d=4 is removed
* Load d=4 → panic, because the node no longer exists

This usually isn't a problem, because the replayed in-memory versions will be the same as before (since it's the same blocks). However, if the pruning settings have been changed, then these orphans can get removed before the change that orphaned them is made, causing this crash. There may be other corner-cases as well.

Will begin work on a fix.

**Comment by alexanderbez** [comment]
What is the proposed fix @erikgrinaker?

**Comment by erikgrinaker** [comment]
I don't know yet. We'll somehow have to avoid persisting the snapshot orphans to disk until a version is actually saved to disk. However, we can't use the recentBatch, since we keep fewer versions in memory than the interval between persisted versions, which means the recentDB has its own orphaning and pruning schedule.

One quick-fix may be to maintain a separate batch for snapshot orphans that haven't been persisted yet. But I feel like the correct fix would be to change the pruning logic to remove `KeepRecent`, and to always keep all version in memory since the last persisted version - that would also simplify a lot of the logic, and avoid many edge-cases.

I will look closer tomorrow.

**Comment by erikgrinaker** [comment]
This was "fixed" by #274.
